### PR TITLE
Let VertexAPI use simplified property serializer

### PR DESCRIPTION
**Pull request by Linary**
Implement #307

Change-Id: I7e47b75160ff240157aec30c1e9b4690a3412e4e

### CODE REVIEW

File: hugegraph-core/src/main/java/com/baidu/hugegraph/io/HugeGraphSONModule.java [hunk 1]
```diff
@@ -118,6 +123,54 @@ private HugeGraphSONModule() {
...
+                jsonGenerator.writeNumber(id.asLong());
+            } else {
+                jsonGenerator.writeString(id.asString());
+            }
+            jsonGenerator.writeStringField("label", vertex.label());
+            jsonGenerator.writeStringField("type", "vertex");
+            // Start write properties
+            jsonGenerator.writeFieldName("properties");
+            jsonGenerator.writeStartObject();
+            vertex.properties().forEachRemaining(vp -> {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/io/HugeGraphSONModule.java]: use `for vp : vertex.getProperties().values()`

File: hugegraph-core/src/main/java/com/baidu/hugegraph/io/HugeGraphSONModule.java [hunk 2]
```diff
@@ -118,6 +123,54 @@ private HugeGraphSONModule() {
...
+                String key = vp.key();
+                Object val = vp.value();
+                try {
+                    jsonGenerator.writeFieldName(key);
+                    JsonSerializer<Object> serializer =
+                            provider.findValueSerializer(val.getClass());
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/io/HugeGraphSONModule.java]: check val is null

File: hugegraph-core/src/main/java/com/baidu/hugegraph/util/JsonUtil.java [hunk 3]
```diff
@@ -44,6 +55,16 @@
...
         module.addSerializer(Date.class, new DateSerializer());
         module.addDeserializer(Date.class, new DateDeserializer());
+
+        module.addSerializer(IdGenerator.StringId.class,
+                             new IdSerializer<>(IdGenerator.StringId.class));
+        module.addSerializer(IdGenerator.LongId.class,
+                             new IdSerializer<>(IdGenerator.LongId.class));
+        module.addSerializer(EdgeId.class, new IdSerializer<>(EdgeId.class));
+
+        module.addSerializer(HugeVertex.class, new HugeVertexSerializer());
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/util/JsonUtil.java]: To ensure that the format of writeObject()  is the same as writeIterator(), also register the serializer to GraphSONWriter.

File: hugegraph-core/src/main/java/com/baidu/hugegraph/io/HugeGraphSONModule.java [hunk 4]
```diff
@@ -276,12 +289,116 @@ public IndexLabelSerializer() {
...
                               JsonGenerator jsonGenerator,
-                              SerializerProvider serializer)
+                              SerializerProvider provider)
                               throws IOException {
             writeEntry(jsonGenerator, schemaSerializer.writeIndexLabel(il));
         }
     }
 
+    private static abstract class HugeElementSerializer<T extends HugeElement>
+            extends StdSerializer<T> {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/io/HugeGraphSONModule.java]: align with class

File: hugegraph-api/src/main/java/com/baidu/hugegraph/serializer/JsonSerializer.java [hunk 5]
```diff
@@ -76,7 +77,8 @@ private String writeList(String label, List<?> list) {
         }
     }
 
-    private String writeList(String label, Iterator<?> itor, boolean paging) {
+    private String writeIterator(String label, Iterator<?> itor,
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/serializer/JsonSerializer.java]: also update writeList(String label, List<?> list) to call JsonUtil.toJson(), keep writeList()

File: hugegraph-api/src/main/java/com/baidu/hugegraph/api/job/GremlinAPI.java [hunk 6]
```diff
@@ -236,11 +252,30 @@ public void checkCreate(boolean isBatch) {
...
+            temp.put("bindings", this.bindings);
+            temp.put("language", this.language);
+            temp.put("aliases", this.aliases);
+            return JsonUtil.toJson(temp);
         }
 
         public static GremlinRequest fromJson(String json) {
-            return JsonUtil.fromJson(json, GremlinRequest.class);
+            @SuppressWarnings("unchecked")
+            Map<String, Object> temp = JsonUtil.fromJson(json, Map.class);
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/api/job/GremlinAPI.java]: rename temp to map

File: hugegraph-api/src/main/java/com/baidu/hugegraph/api/job/GremlinAPI.java [hunk 7]
```diff
@@ -236,11 +252,30 @@ public void checkCreate(boolean isBatch) {
         }
 
         public String toJson() {
-            return JsonUtil.toJson(this);
+            Map<String, Object> temp = new HashMap<>();
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/api/job/GremlinAPI.java]: rename temp to map

File: hugegraph-core/src/main/java/com/baidu/hugegraph/io/HugeGraphSONModule.java [hunk 8]
```diff
@@ -276,21 +289,125 @@ public IndexLabelSerializer() {
...
                               JsonGenerator jsonGenerator,
-                              SerializerProvider serializer)
+                              SerializerProvider provider)
                               throws IOException {
             writeEntry(jsonGenerator, schemaSerializer.writeIndexLabel(il));
         }
     }
 
-    private static class ShardSerializer extends StdSerializer<Shard> {
+    private static abstract class HugeElementSerializer<T extends HugeElement>
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/io/HugeGraphSONModule.java]: set protected

File: hugegraph-api/src/main/java/com/baidu/hugegraph/metric/MetricsModule.java [hunk 9]
```diff
@@ -0,0 +1,251 @@
+package com.baidu.hugegraph.metric;
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/metric/MetricsModule.java]: add apache2 license

File: hugegraph-test/src/main/java/com/baidu/hugegraph/unit/core/JsonUtilTest.java [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+            return;
+        }
+
+        try {
+            graph.clearBackend();
+        } finally {
+            try {
+                graph.close();
+            } catch (Throwable ignored) {
+                // pass
```
Comment (zhoney) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/unit/core/JsonUtilTest.java]: // ignore

File: hugegraph-test/src/main/java/com/baidu/hugegraph/unit/core/JsonUtilTest.java [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+            }
+            graph = null;
+        }
+    }
+
+    @Before
+    public void setup() {
+        // pass
+
```
Comment (zhoney) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/unit/core/JsonUtilTest.java]: delete empty line

File: hugegraph-test/src/main/java/com/baidu/hugegraph/unit/core/JsonUtilTest.java [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+    private static HugeGraph graph = null;
+
+    @BeforeClass
+    public static void initEnv() {
+        RegisterUtil.registerBackends();
+    }
+
+    @BeforeClass
+    public static void init() {
+        graph = Utils.open();
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/unit/core/JsonUtilTest.java]: just use mock

**Pull request merged**